File: frontend/src/services/category.service.ts
import axios from "../config/axios.config";
import type { ApiResponse, ErrorApiResponse } from "../types/api.type";
import type { Category, RawCategory, CategoryFormData } from "../types/category.types";

const BASE_URL = "/categories";

export const CategoryService = {
  async getAll(): Promise<ApiResponse<Category[]>> {
    try {
      const res = await axios.get(BASE_URL);

      const rawData: RawCategory[] = Array.isArray(res.data)
        ? res.data 
        : Array.isArray(res.data.data)
        ? res.data.data 
        : [];

      const mapped: Category[] = rawData.map((c) => ({
        id: c.categoryId,
        name: c.categoryName,
      }));

      return {
        success: true,
        message: "Lấy danh mục thành công",
        statusCode: 200,
        data: mapped,
      };
    } catch (error: any) {
      const err = error as ErrorApiResponse;
      return {
        success: false,
        message: err.message ?? "Lỗi không xác định",
        statusCode: err.statusCode ?? 500,
        data: [],
      };
    }
  },

  async create(payload: CategoryFormData): Promise<ApiResponse<Category>> {
    try {
      const res = await axios.post<RawCategory | ApiResponse<RawCategory>>(
        BASE_URL,
        { categoryName: payload.name },
        { headers: { "Content-Type": "application/json" } } 
      );

      const c: RawCategory =
        (res.data as any).data ?? (res.data as RawCategory);

      const mapped: Category = {
        id: c.categoryId,
        name: c.categoryName,
      };

      return {
        success: true,
        message: "Tạo danh mục thành công",
        statusCode: 200,
        data: mapped,
      };
    } catch (error: any) {
      const err = error as ErrorApiResponse;
      return {
        success: false,
        message: err.message ?? "Lỗi tạo danh mục",
        statusCode: err.statusCode ?? 500,
      };
    }
  },

  async update(id: number, payload: CategoryFormData): Promise<ApiResponse<Category>> {
    try {
      const res = await axios.put<RawCategory | ApiResponse<RawCategory>>(
        `${BASE_URL}/${id}`,
        { categoryName: payload.name },
        { headers: { "Content-Type": "application/json" } }
      );

      const c: RawCategory =
        (res.data as any).data ?? (res.data as RawCategory);

      const mapped: Category = {
        id: c.categoryId,
        name: c.categoryName,
      };

      return {
        success: true,
        message: "Cập nhật danh mục thành công",
        statusCode: 200,
        data: mapped,
      };
    } catch (error: any) {
      const err = error as ErrorApiResponse;
      return {
        success: false,
        message: err.message ?? "Lỗi cập nhật danh mục",
        statusCode: err.statusCode ?? 500,
      };
    }
  },

  async delete(id: number): Promise<ApiResponse<null>> {
    try {
      await axios.delete(`${BASE_URL}/${id}`);
      return {
        success: true,
        message: "Xoá danh mục thành công",
        statusCode: 200,
        data: null,
      };
    } catch (error: any) {
      const err = error as ErrorApiResponse;
      return {
        success: false,
        message: err.message ?? "Lỗi xoá danh mục",
        statusCode: err.statusCode ?? 500,
        data: null,
      };
    }
  },
};

File: blazor_web/Services/Category/CategoryService.cs
using blazor_web.Models;         
using System.Net.Http.Json;
using System.Text.Json; // Cần thêm namespace này
using System.Text.Json.Serialization; 

namespace blazor_web.Services.Category;

public class CategoryService
{
    private readonly HttpClient _http;

    public CategoryService(HttpClient http) => _http = http;

    // 1. GET ALL
    public async Task<ApiResponse<List<CategoryResponse>>> GetAll()
    {
        try
        {
            var data = await _http.GetFromJsonAsync<List<CategoryResponse>>("categories");

            return new ApiResponse<List<CategoryResponse>>
            {
                Success = true,
                Message = "Lấy dữ liệu thành công",
                Data = data ?? new List<CategoryResponse>()
            };
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error: {ex.Message}");
            return new ApiResponse<List<CategoryResponse>>
            {
                Success = false,
                Message = "Lỗi kết nối hoặc sai định dạng dữ liệu",
                Data = new List<CategoryResponse>()
            };
        }
    }

    // 2. CREATE (Đã nâng cấp bắt lỗi)
    public async Task<ApiResponse<CategoryResponse>> Create(CategoryCreateRequest req)
    {
        try
        {
            var response = await _http.PostAsJsonAsync("categories", req);
            
            if (response.IsSuccessStatusCode)
            {
                var data = await response.Content.ReadFromJsonAsync<CategoryResponse>();
                return new ApiResponse<CategoryResponse> { Success = true, Data = data! };
            }

            // SỬA: Đọc lỗi chi tiết từ Backend
            string errorMsg = await GetErrorMessageAsync(response, "Tạo thất bại");
            return new ApiResponse<CategoryResponse> { Success = false, Message = errorMsg };
        }
        catch (Exception ex)
        {
            return new ApiResponse<CategoryResponse> { Success = false, Message = $"Lỗi kết nối: {ex.Message}" };
        }
    }

    // 3. UPDATE (Đã nâng cấp bắt lỗi)
    public async Task<ApiResponse<CategoryResponse>> Update(int id, CategoryCreateRequest req)
    {
        try
        {
            var response = await _http.PutAsJsonAsync($"categories/{id}", req);
            
            if (response.IsSuccessStatusCode)
            {
                var data = await response.Content.ReadFromJsonAsync<CategoryResponse>();
                return new ApiResponse<CategoryResponse> { Success = true, Data = data! };
            }

            // SỬA: Đọc lỗi chi tiết từ Backend
            string errorMsg = await GetErrorMessageAsync(response, "Cập nhật thất bại");
            return new ApiResponse<CategoryResponse> { Success = false, Message = errorMsg };
        }
        catch (Exception ex)
        {
             return new ApiResponse<CategoryResponse> { Success = false, Message = $"Lỗi kết nối: {ex.Message}" };
        }
    }

    // 4. DELETE (Đã nâng cấp bắt lỗi)
    public async Task<ApiResponse<object>> Delete(int id)
    {
        try
        {
            var response = await _http.DeleteAsync($"categories/{id}");
            
            if (response.IsSuccessStatusCode)
            {
                return new ApiResponse<object> { Success = true, Message = "Xóa thành công" };
            }

            // SỬA: Đọc lỗi chi tiết từ Backend
            string errorMsg = await GetErrorMessageAsync(response, "Xóa thất bại");
            return new ApiResponse<object> { Success = false, Message = errorMsg };
        }
        catch (Exception ex)
        {
             return new ApiResponse<object> { Success = false, Message = $"Lỗi kết nối: {ex.Message}" };
        }
    }

    // ================= HÀM PHỤ ĐỂ ĐỌC LỖI TỪ BACKEND =================
    private async Task<string> GetErrorMessageAsync(HttpResponseMessage response, string defaultMessage)
    {
        try
        {
            // Đọc toàn bộ nội dung trả về
            var content = await response.Content.ReadAsStringAsync();

            // Nếu nội dung rỗng, trả về mặc định
            if (string.IsNullOrWhiteSpace(content)) return defaultMessage;

            // Thử parse xem có phải JSON format { "message": "..." } không (Format backend bạn đang dùng)
            var options = new JsonSerializerOptions { PropertyNameCaseInsensitive = true };
            var errorDto = JsonSerializer.Deserialize<BackendErrorDto>(content, options);

            if (!string.IsNullOrEmpty(errorDto?.Message))
            {
                return errorDto.Message; // Trả về thông báo cụ thể từ Backend
            }
            
            // Nếu không phải JSON hoặc không có field message, trả về raw content (nếu ngắn)
            if (content.Length < 200) return content;
        }
        catch
        {
            // Nếu lỗi parse JSON, bỏ qua
        }

        return $"{defaultMessage} (Mã lỗi: {response.StatusCode})";
    }
}

// Class phụ để hứng lỗi JSON từ Backend
public class BackendErrorDto
{
    public string Message { get; set; } = string.Empty;
}

// ================= DTO =================

public class CategoryCreateRequest
{
    [JsonPropertyName("CategoryName")] 
    public string Name { get; set; } = string.Empty;
}

public class CategoryResponse
{
    [JsonPropertyName("CategoryId")] 
    public int Id { get; set; }

    [JsonPropertyName("CategoryName")] 
    public string Name { get; set; } = string.Empty;
}